### ISSUE TITLE
[BUG] chunkserver: catch ext4 read-only state

### ISSUE THREAD

**Issue by onlyjob** [moosefs/moosefs#452]
While number of chunks on chunkserver did not change, I've noticed unrealistically high number of ongoing chunks deletions (13k) in chart, which reminded me of #273 (don't make fake charts by visualising unfulfilled intentions. Fictitious charts are useless).

Quick look into logs revealed that a disk marked for removal developed errors and `ext4` put it in _read-only_ state (which I've managed to clear later by `fsck.ext4`).
However Chunkserver continued to try deleting chunks without proper handling of this condition or raising "damaged" state on read-only mount. Perhaps such condition should be detected and handled?

Observed on 3.0.116.

**Comment by chogata** [comment]
What you describe here doesn't make much sense...
- if a disk is marked as MFR, master doesn't send requests for chunk deletions for chunks that are on this disk (this is the only case when master actually knows and cares on which disk on a chunk server the chunk is recorded, in that it remembers if it is on MFR or not)
- MooseFS makes statvfs call every minute and if the disk state changes from rw to ro, it is marked as damaged.

BUT... if a disk is already in a read only state when the chunk server starts, it will be marked as MFR (for master's sake) and with a special ro flag in the chunk server itself, that will prevent the disk from being marked as damaged, unless the ENOENT error occurs (a file went missing from the disk). So what you describe (no state change to damaged) is only possible, if you mark a disk as MFR and restart the chunk server and this is the moment when the disk goes ro and you don't notice. MooseFS CGI will not show this special ro flag, so there is no way to know. You would notice, if you did not mark it for removal, as then the system would do it for you.

But that's only a possible explanation for the "not damaged" part of your issue. As I said, there should be no deletions send for chunks on MFR disks. I can't really imagine why unsuccessful replications would be send to other disks at the same time. If you have any more information on the situation that might help, please share it.

And on a completely side note: unsuccessful attempts to delete/replicate something have to be reported, otherwise if the system is "not replicating" or "not deleting" something we don't know if this is because the master doesn't deem it necessary to order an operation be done, or because the operation fails somehow. This is the first step to diagnose a problem. And we split the charts in two shaded some time ago, to adequately present how many successful and unsuccessful operations of certain type happened.

**Comment by onlyjob** [comment]
MooseFS removes chunks from disks marked for removal just no as aggressively as on normal disks.
I observed that many times in CGI where counter or chunks on MFR disk slowly goes down. Perhaps you know better what is causing that.
Issue should be easy enough to reproduce by re-mounting disk as read-only on the fly, while chunkserver is running.

**Comment by chogata** [comment]
Okay, you are right, MFS does delete chunks on MFR disks, but only if/when files that those chunks belong to are deleted from the filesystem. I forgot to add that. But that cannot have been your problem, unless you suddenly deleted a lot of files from your instance when all that happened.
We can try to re-create the situation with the MFR disk going read-only, but I doubt we will get those chunk deletions without replying *exactly* the same operations that happened on your system then. And we don't know what they were (unless you tell us :) ).

**Comment by onlyjob** [comment]
A lot of thing is constantly going on in the cluster. CI jobs are running and large number of deletions are likely to happen at any time. There are 30_000 to 60_000 files in trash at any given time (60...100 GiB or more.) I don't know about all operations in the cluster at a time but I did my best to report observations accurately.

**Comment by chogata** [comment]
OK, if your system is constantly deleting files, then those deletions could have been "regular file deletion" deletions. It's a bit of a catch 22 situation with a read-only MFR, but we'll think if we can find a way around it.

"There is also a possibility that HDD was tipped over into read-only state by large number of deletions." - nope, that was not possible. If the ro state happened on the disk after CS was restarted, it would have been marked as damaged. The only way to not mark a disk as damaged is when it is ro at CS process start. Otherwise any 2 consecutive i/o errors would trigger the damaged condition (as opposed to ENOENT error only for ro disks).

**Comment by onlyjob** [comment]
Interesting. So I might have restarted chunkserver before realising that HDD marked for removal is already mounted read-only?

**Comment by chogata** [comment]
Exactly. To be precise: it had to change to read only sometime in between you giving the chunkserver the restart command and the moment the chunkserver started to scan this disk after starting again.
